### PR TITLE
fix: remove options structs in go sdk

### PR DESCRIPTION
**Pull request by jedevc**
:warning: Based on https://github.com/dagger/dagger/pull/5893
:hammer_and_wrench: Fixes https://github.com/dagger/dagger/issues/5904 https://github.com/dagger/dagger/issues/5864

This PR uses the approach detailed in https://github.com/dagger/dagger/issues/5864#issuecomment-1765040417:

> To resolve this, instead, we can just look at the Name of the struct - if it ends in "Opt" or similar, then we treat it as an opts struct.

All structs that end in "Opt"/"Opts"/"Option"/"Options" are treated as an opts-like struct, and we read parameters directly into them. However, because we're now not using the pointeriness of parameters to determine this, there's now a *bunch* of cases we need to handle - how should we handle `**Opts` or `[]Opts`, etc. These are all handled now by creating a single opts struct, and then working on how to put it into the target type.

### CODE REVIEW

**Comment by jedevc** [comment]
One known case that this *doesn't* cover:

```go
type EchoOpts struct {
	Repeat int
}

type NestedOpts struct {
	Echo EchoOpts
}

func (m *Minimal) EchoOptsNested(msg string, opts NestedOpts) string {
	...
}
```

Not quite sure what the right thing in this case to do would be, but I would assume it should flatten everything out (so EchoOpts also gets filled from the args). This would be quite a useful pattern to allow extending groups of args.

**Comment by sipsma** [comment]
I have comments+questions on the discussion in the related issue ([here](https://github.com/dagger/dagger/issues/5864#issuecomment-1765040417)), leaving them here since it gets intertwined with the actual implementation.

1. Relying on a suffix for type names to get certain behavior doesn't feel great
   * One thing we'd like to support in the longer term is the ability to import structs outside the package (even from 3rd party libs) and use those as inputs/outputs. Relying on them having certain naming patterns (or requiring the user wrap/alias to get the desired naming pattern) feels like a significant paper cut
   * The documentation overhead, not just of writing but for docs that users need to read, is also significantly expanded here
2. I'm sort of confused around the behavior when there's a `[]FooOpt`
   * In [this test case](https://github.com/dagger/dagger/pull/5907/files#diff-3f816215b584a031057b440089ceb90784c3d4257816d7facfa3dcec14fe0a94R559), it seems like even though there's a list of opts being accepted, we can only fill out the fields of one? Does that mean we can only input a slice with a single element? Or am I misunderstanding?

---

Overall, I'm kind of wondering whether we should take the current sort of Opt approach at all and instead find different ways of solving the problems it was solving. I'm imagining trying to write the docs for all this behavior and the result being very confusing for users. If I, as a user, accept a list of `[]FooOpt`, I should be able to provide a slice of values of that struct when calling it; it's hugely surprising if that's not the case.

Am I correct that this is what `Opt`s currently solve?
1. Ability to distinguish between required+optional parameters in Go
2. Ability to add doc strings for input parameters in Go

If so, I have some ideas on different approaches that could maybe avoid the need to support the `Opt` pattern and achieve those in a different way, but I'll wait for confirmation I'm anywhere near on the same page as everyone else before expanding my current wall-of-text further :-)

cc @vito @jedevc

**Comment by jedevc** [comment]
> Relying on a suffix for type names to get certain behavior doesn't feel great 

Agreed - in Python, you'd be able to add a decorator, which would be awesome. But in go, you can't and so it feels a bit stuck. Tagging each field in each struct also doesn't feel quite right, since opts-ishness is a property of the struct, not fields in that struct (or the type of that struct in args).

> One thing we'd like to support in the longer term is the ability to import structs outside the package (even from 3rd party libs) and use those as inputs/outputs

Agreed! However, I don't think that we want arbitrary third-party structs used as opts-like structs - it's more likely useful to have these used as IDable types.

> The documentation overhead, not just of writing in terms of what users need to read, is also significantly expanded here

I don't think this is any harder to document than the current pointer behaviour determining whether something is an opt.

> I'm sort of confused around the behavior when there's a `[]FooOpt`

This is because there's only ever conceptually a single opt passed to a function. It's more a question of completeness - what should we as codegen do when we see a type of `[]OptionyType`? There's a few options:

- Pretend it's not an optiony type, and that it's IDable - this isn't true, so not really a good option IMO.
- Error out and indicate the user that this is a weird pattern.
- Just accept only one option and find a way of passing it in (what we do currently) and maybe warn the user that they shouldn't do it.

These are all kinda weird, but I'd lean into "works but warn".

The idea is that we unpack all the opt fields to the top-level to do a couple of things:

> Am I correct that the following is what Opts currently solve?

- Allow bundling params together, so you don't have lots and lots of individual args (and also potentially allow commonality between different functions).
- Allow marking args as optional.
- Allow documenting args.

> If so, I have some ideas on different approaches that could maybe avoid the need to support the Opt pattern

Me and @vito also discussed some other options:

- Use a code-gened generic type like `Embed[T]` - if an arg has this value, then do the embedding behaviour.
  - Pros: explicit
  - Cons: requires generics, isn't a property of the struct, may be a little fiddly for users to use, still requires that we answer "what do we do when `T` is a pointer, or a slice, etc"
- Opts are indicated by using variadic args - all variadic args are treated as opts
  - Pros: still explicit, fits the common go idiom
  - Cons: requires the user handles *multiple* opts structs (even if we call with just one, the module could call itself with multiple), and still isn't a property of the struct itself

To be fair, having implemented this whole thing, I'm now slightly more convinced to give `Embed` a try - I like it better than the name inspection.

If we could find a way to avoid `Opt` entirely, I could definitely be convinced,

**Comment by vito** [comment]
While we're on this topic, another elephant in the room is how to annotate required args with documentation. I've in fact relied on inline structs + `required:"true"` and `default:"foo"` previously to define full argument specs before:

```go
func (Apko) Alpine(args struct {
	Packages []string `doc:"List of package names to install." required:"true"`
	Branch   string   `doc:"Alpine branch to use." default:"edge"`
}) (*Container, error) {
	ic := baseConfig()
	ic["contents"] = cfg{
		"repositories": []string{
			"https://dl-cdn.alpinelinux.org/alpine/" + args.Branch + "/main",
		},
		"packages": append([]string{"alpine-base"}, args.Packages...),
	}
	return apko(ic)
}
```

I completely forgot about this when suggesting that we rely on `Opts` to deduce whether it's an options struct.

Once you get past the somewhat novel use of an inline struct type, it's actually quite nice: all of your arguments, their documentation, their defaults, and their required-ness are all in one place, making it very easy to rapidly put together APIs that look great from the outside and don't require any throw-away/one-time-use types.

The downside of this pattern is that intra-module calls are basically not supported anymore, but really, intra-module calls are already kind of broken because they circumvent all our caching special sauce. I tried to nerd-snipe @jedevc into supporting some sort of `dag.Self()` for that. :wink:

TBH the best times I've had building CLIs has been when they heavily lean on struct tags like above. I was a big fan of [jessevdk/go-flags](https://pkg.go.dev/github.com/jessevdk/go-flags) for example.

So I guess what I'm wondering is if there's a path we can take where it's _all_ just one big struct definition, instead of relying on naming, worrying about `[]FooOpts`, etc. etc.

As a side note, I agree that relying on naming isn't ideal, but on the other hand it's not exactly a novel pattern, enforcing conventions is not a bad thing, and it doesn't seem that hard to document. We're also really limited by Go's expressiveness here, so it's mostly about finding the "least bad" solution.

**Comment by sipsma** [comment]
> for required/optional args, we could potentially have a generic Optional[T] that would just be a type alias? 

Yeah that's what I was thinking, the go sdk includes a type like:
```go
type Optional[T any] struct {
  value T
  isSet bool
}

func (o Optional[T]) Value() (T, bool) {
  return o.value, o.isSet
}
```

Which is then used like:
```go
func (myMod *MyMod) MyFunc(arg Optional[string]) error {
  argVal, ok := arg.Value()
  // ...
}
```

IIRC, the `x/tools/types` package we use in the go codegen has support for identifying generic types, the type of the parameter, etc. So it shouldn't be that much complication for us. And for users I think it's fairly obvious.

I like this approach because it's totally orthogonal with other concerns (like doc strings or whether or not certain params are bundled into a struct or not) and only becomes something users need to be aware of once they really care about it (the default path of just writing a normal looking go method that results in required args is a perfectly fine default imo).

> For docs, we could try and follow a standarized doc-comment convention

Yeah, we talked about that in the past because at the time it was the only thing I could think of. But tbh the idea of essentially creating our own version of JSDoc but for Go is really unappealing. It seems like a huge gap in go docs that there isn't a standard for this, but that's not a problem we're gonna solve unless we want to take months/years trying to upstream something.

However, @vito had a great idea a while back of something like this:
```go
func (myMod *MyMod) MyFunc(
  // foo is some sort of string that I can doc in this comment
  foo string,
  // bar is some sort of Container that I can doc in this comment
  bar *Container,
) error {
  // ...
}
```

At the time we ruled it out because the go ast library doesn't have builtin support for "give me the comment above this parameter" (as opposed to comments above methods, struct fields, etc. where there is builtin support since it's needed for go doc generation).

However, that was back in the old version of the Go SDK which was doing a weird mix of reflection+ast parsing and using the low-level ast library directly (as opposed to the `x/tools/types` package that we use now). It seemed very hard due to multiple factors, but the biggest being that since it was at runtime and using reflection in part, trying to pinpoint the comment above a given arg started contending with go compiler optimizations that cause line numbers to no longer be accurate (😵‍💫).

I think it's worth revisiting now that we are doing pure AST parsing and have a better library to support us. I can't think of a reason it would be too herculean a task to:
1. Get the line number of a function arg
2. Look at the line numbers above and see if they are comments, use them as doc strings if so

Of course, that theory may not work out in practice, but if it works I think it would be a pretty natural and very easy-to-doc approach. We just include an example of it in our docs and it should click right away in users' minds (I think).

> But I don't see any good way to solve the bundling one without an Opts struct, if that's something we need to keep.

We definitely need to support structs as inputs. The idea pre-my-vacation was just to make them be graphql Input types and do the normal json marshal/unmarshal to send them over the wire, with extra handling for the corner case where a type is used as both input and output (need two type definitions, which while annoying would be entirely handled server-side in the engine, so not something anyone needs to think hard about in terms of SDKs).

That still is a viable option, but is contradictory with our efforts to make everything ID-able, in which case we'd have the graphql schema accept an ID rather than graphql input type. Maybe it's okay as an intermediate option still while we work towards the universal ID-able approach? Provided we can update without it being too painful a breaking change.

I have more thoughts on this, but I'll do another checkpoint and get others input before continuing 🙂

**Comment by vito** [comment]
I like `Optional[T]` too, though it still leaves a gap wherein you can't specify a default value in a way that's easy for the SDK to detect and hoist up to the API level.

Basically my end goal is to support rich docs like this in Daggerverse:

image

(Please ignore the misalignment and janky colors, this was hand-edited in Firefox Inspector and I give up. :cry:)

The nice thing about the `args struct {}` approach is that it handles all that, and also means you don't have to write any boilerplate to assign default values when optional arguments are not set.

It also lets you specify a 'required with a default' argument, which is a subtly different thing from a truly-optional argument, and is often what you want instead.

And if we can get doc comments working too, that's great - though I'm a little sketched out by the fact that Go linters can be very opinionated about how those are written ("Package foo does bar, Foo does this") in a way that doesn't really match GraphQL conventions. I think we're OK when it comes to function args and struct fields though.

**Comment by sipsma** [comment]
> The nice thing about the args struct {} approach is that it handles all that, and also means you don't have to write any boilerplate to assign default values when optional arguments are not set.
It also lets you specify a 'required with a default' argument, which is a subtly different thing from a truly-optional argument, and is often what you want instead.

I do see the appeal, but while encoding the default value in a ``` `tag` ``` makes total sense for primitive types, once you get to lists, structs and especially ID-able types like Container/Directory, I think it starts to make less sense.

I think that would be a real use-case too; e.g. my function accepts a base image to use as an argument, but I want there to be a default value of some container.

The only way I can see of supporting that would be to put the container ID as the default value, which while technically possible, would not be great DX of course. Let me know if you have any other ideas though.

---

For the sake of an alternative suggestion:

```go
type MyStrArg string

func (MyStrArg) Default() string {
	return "da default"
}

type MyContainerArg Container

func (*MyContainerArg) Default() *Container {
	return dag.Container().From("blah").WithExec(...)
}

func (*MyModule) MyFunction(strArg MyStrArg, ctr *MyContainerArg) error {
  // ...
}
```

Basically, check if the arg type implements `interface[T any] { Default() T }`, with `T` set to the underlying type of the named type.
* We may need multiple interfaces to check against (internally, not user-facing concern) so we could allow `(T, error)` as return too, but same basic idea applies.

This is verbose, don't love it, but it would get us the full power of arbitrary default values, so it's seems worth putting on the table.

---

Putting everything together, a Function could look like:
```go
type MyContainerArg Container

func (*MyContainerArg) Default() *Container {
	return dag.Container().From("blah").WithExec(...)
}

func (*MyModule) MyFunction(
  // docs for strArg
  strArg string,
  // docs for intArg
  intArg Optional[int],
  // docs for baseImage
  baseImage *MyContainerArg,
) error {
  // ...
}
```

Personally, I think I'm okay with that. The only thing I don't like is the boilerplate of default arg values, but it also trades off boilerplate of inline structs + ``` `tag` ```s. And we'd not be limiting ourselves in terms of what types can have default values and what can't.

I think what I like the most overall about this approach is that it retains the ability for users to still just write extremely familiar code like this:

```go
func (*MyModule) MyFunction(strArg string,  intArg int,  baseImage *MyContainerArg) error {
  // ...
}
```

That would still be totally valid and "good-enough" code to start, which I think keeps the barrier of entry and surprise-factor low. You can start there, with go code that looks pretty much the same as what you are familiar with, then as you want to refine further you can delve into our docs further and learn how to make something optional, give something a default value, etc.

**Comment by jedevc** [comment]
Ok, so to summarize so far: it seems like we're all on the same page that 1. we shouldn't detect opts using pointers, and 2. we shouldn't detect opts using naming conventions.

Long-term it looks like we want to make all struct types IDable, but we still need a way to have optional/required, doc-strings and default values for parameters.

There are two main ideas:

1. Use functions, insert doc-strings between line-separated parameters, and have an `Optional[T]` to indicate that a parameter is optional
    - Unclear what we should do for defaults
3. Use an inline struct (moving all field names up a level to the parameter), use struct doc-comments, and use field tags to indicate defaults and optional (or use `Optional[T]`)
    - Intra-module calls are no fun
    - Default tags can't capture the variety of defaults we might want

---

> ```go
> func (myMod *MyMod) MyFunc(
>   // foo is some sort of string that I can doc in this comment
>   foo string,
>   // bar is some sort of Container that I can doc in this comment
>   bar *Container,
> ) error {
>   // ...
> }
> ```

I actually quite like this. It hadn't occurred to me that you were allowed comments inside the args declaration. It doesn't seem *too* bad to try and hack this together, even if the go parser doesn't help us out here. It's definitely better than rolling our own standard way to document args.

> I like `Optional[T]` too

Awesome, we seem to have consensus on liking `Optional[T]` then :tada:

It reminds me a bit of [FastAPI](https://fastapi.tiangolo.com/python-types/#python-types-intro) which generates OpenAPI schemas from the types of parameters, and checks them as a basic form of validation.

Funnily enough - they have an equivalent "opts"-like struct, if you create a class that extends `BaseModel`, then the query parameters are unpacked into that class. Maybe if we wanted to keep the opts pattern at all (though with the above discussion, I'm kind of convinced that we could get away with not having it) we could do so with a similar extension pattern - if you embed a special `Model` struct/interface, then we know to unpack to that level (essentially this marks an opts struct as "opts-like").

>> though it still leaves a gap wherein you can't specify a default value
>
> I do see the appeal, but while encoding the default value in a `tag` makes total sense for primitive types, once you get to lists, structs and especially ID-able types like Container/Directory, I think it starts to make less sense

I'm not actually sure what we should do for default IDable types. What would that look like in daggerverse? I imagine that the way the default would work would be using an ID, but I'm not quite sure what it should *look* like in docs / when importing to another language.

For structs, I think I'd like to move to a world in which *all* structs are IDable (unless we do a `BaseModel` like approach from above, or similar). So that case is the same as above.

But agreed, even imagining a list of lists is quite challenging in a `default` tag, what should that even look like?

> Basically, check if the arg type implements `interface[T any] { Default() T }`, with `T` set to the underlying type of the named type.

Agh, I'm really struggling to come up with anything that I like better than this, but this feels *super* unwieldy - it means I need a new type for every new default.

In a beautiful, ideal world, we could have something like `Default[int, 100]`, where we could pass constant values into parameters (even that wouldn't quite work though, since slices/maps aren't constants lol).

The problem is, whatever we do is gonna be hacking around go's lack of default params - the go way is more to set the default from within the function if we really need parameters like that.

Maybe this is worth a brainstorming session, but here's a couple other bizarre ideas:

- In a specially named function, we call all the other functions that we want to have defaults - the values from that function call are made the defaults for it.
- We could potentially actually just run the function and capture the values of the parameters (and set those as the defaults) - this is truly insane though and just asking for problems.

**Comment by vito** [comment]
> I do see the appeal, but while encoding the default value in a `` `tag` `` makes total sense for primitive types, once you get to lists, structs and especially ID-able types like Container/Directory, I think it starts to make less sense.

Good point. The struct tags approach already supports JSON-encoded values, and technically supports literal IDs but you'd obviously never want to type them out.

* Sidebar: if we're going to support ID-able default values, we're going to need a sensible way to represent those in Daggerverse. With v2 IDs (#3923) I could unpack the ID and render its inner query chain, which is kind of neat. Something like `container.from(address: "alpine/git")`. Hopefully no one would use anything too complicated as a default. :sweat_smile: 

That same `interface[T any] { Default() T }` pattern could just be applied to the struct fields, though:

```go
type MyContainerArg Container

func (*MyContainerArg) Default() *Container {
	return dag.Container().From("blah").WithExec(...)
}

func (*MyModule) MyFunction(args struct {
  StrArg string          `doc:"String arg with a default." default:"da default"`
  Ctr    *MyContainerArg `doc:"Container arg with a default."`
}) error {
  // ...
```

Or defaults could be pulled out to vars and referenced:

```go
// GitUtil provides various utilities for working with Git repositories.
type GitUtil struct{}

// DefaultBase is the default base image used for git commands.
var DefaultBase = dag.Apko().Wolfi([]string{"git"})

// WithRepo sets the repo for future calls to run against.
func (m *GitUtil) Repo(args struct {
	URL  string     `doc:"URL of the git repository." required:"true"`
	Base *Container `doc:"Base image to use for git commands." defaultVar:"DefaultBase"`
}) *GitRepo {
	return &GitRepo{
		Base: args.Base,
		URL:  args.URL,
	}
}
```

Still obviously assumes you're OK with the inline struct pattern, but it's a whole lot less to type at least.

---

> I think what I like the most overall about this approach is that it retains the ability for users to still just write extremely familiar code like this:

I fully recognize this pattern looks a little odd if you're comparing to regular old Go functions, but so does a lot of other glue code like CLI frameworks. If I were to compare this model the alternative, I would argue that the inline struct approach is more optimized for pumping out a ton of glue code with high quality APIs and minimal boilerplate, Go conventions be damned, because this isn't designed to be called by Go code.

But the real beauty is that if no one else likes it I can just make it my own SDK. :grin: 

* Sidebar: When the issue of ID-able defaults came up in the past I went with a [`withFoo` pattern](https://github.com/vito/dagger-gitutil/blob/810a6972dd8556fe6b341036aca119a7d85995f7/main.go#L13-L53) instead, which completely sidesteps the issue since there's never actually a default value for an argument anywhere - it's just handled at runtime instead. It'll be interesting to see if this was invention by necessity, since defaults were unsupported, or if this pattern will continue.

**Comment by jedevc** [comment]
Me, @vito and @sipsma had a brief sync, and made some decisions, that I'll try and start implementing tomorrow:

- We're going to support inline struct opts, and native function params
- All named structs will be treated as IDable (even if they're currently not, the follow-up will be to actually make them all IDable)
- Docs for parameters will be read as godocs from the inline structs, and magical string inspection for the native function params
- Optional args will be defined with a special type `Optional[T]`
- Default parameters will not be explicitly supported in the Go SDK
  - Users can still do defaults, by writing `if`s in their functions, but this won't propagate to docs in the daggerverse, or codegen in other languages.
  - This is sad, but there's no really obvious way to do it at the moment. Ideally, we don't include this, some user notices and complains and gives us a perfect solution that we haven't thought of :tada:

**Comment by jedevc** [comment]
Ok, I've pushed a patch to get *most* of the tricky bits from the above - the only big missing part is `Optional[T]`, but I've gotten the comments parsing working nicely (which was the thing I was worried if it was possible or not).

Should hopefully not be too much work, but probably don't have enough time to polish it to a level I'd be happy with before the planned release tomorrow (but should definitely make the next one).

**Comment by jedevc** [comment]
*many edge cases later...*

:tada: We have a new `Optional` type! Happy to bikeshed exactly what it should look like, I just gave it some helpers to simplify JSON unmarshalling, and `Get`/`GetOr` functions to get the underlying data. I also realized we need an easy way to construct `Optional`s so we can do easy intra-module calls, so I make `Opt` as a quick constructor.

Otherwise, I think this is pretty much what we discussed last week - I'll keep updating with some tests as I think of more edge cases, but should be ready for review/merge/etc.

File: cmd/codegen/generator/go/templates/modules.go [hunk 1]
```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
...
-				if len(argOpts) > 0 {
-					argArgs = append(argArgs, Id("FunctionWithArgOpts").Values(argOpts...))
-				}
+	fnDecl, err := ps.declForFunc(fn)
+	if err != nil {
+		return nil, err
+	}
 
-				fnDef = dotLine(fnDef, "WithArg").Call(argArgs...)
+	if i+1 == params.Len() {
```
Comment (sipsma) [inline cmd/codegen/generator/go/templates/modules.go:788]: nit: took me a sec to grok, a quick comment about the case this is handling (inline struct) would be helpful

File: cmd/codegen/generator/go/templates/modules.go [hunk 2]
```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
...
+	paramType types.Type
+	// baseType is the simplified base type derived from the function signature
+	baseType types.Type
+
+	optional bool
+	variadic bool
+
+	defaultValue string
+
+	parent *paramSpec
```
Comment (sipsma) [inline cmd/codegen/generator/go/templates/modules.go:898]: nit: quick comment on what `parent` is would help, otherwise have to dig through code to grok the case it's used for

File: cmd/codegen/generator/go/templates/modules.go [hunk 3]
```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
...
+		ptr, ok := baseType.(*types.Pointer)
+		if !ok {
+			break
+		}
+		baseType = ptr.Elem()
+	}
+
+	optional := false
+	if named, ok := baseType.(*types.Named); ok {
+		if named.Obj().Name() == "Optional" {
```
Comment (sipsma) [inline cmd/codegen/generator/go/templates/modules.go:857]: nit: may want to also check if `named` is coming from `dagger.gen.go`, just to future-proof against support we add eventually to use types from external packages (which could also be named `Optional` here since `.Obj().Name()` returns an unqualified name)

File: cmd/codegen/generator/go/templates/modules.go [hunk 4]
```diff
@@ -677,99 +716,190 @@ func (ps *parseState) goMethodToAPIFunctionDef(typeName string, fn *types.Func,
...
+
+	optional bool
+	variadic bool
+
+	defaultValue string
+
+	parent *paramSpec
+}
+
+func (spec *paramSpec) graphqlName() string {
```
Comment (sipsma) [inline cmd/codegen/generator/go/templates/modules.go:901]: nit: after https://github.com/dagger/dagger/pull/5932 the engine always provides object+function names with the same name (including casing) that they were originally provided by the SDK as (i.e. SDK registers function with `CamelCase`, engine will invoke SDK runtime using that too even though internally those get converted to `lowerCamelCase` for the gql schema).

So I wouldn't have thought we'd need to handle anything like this anymore. But may be missing something. Not a big deal, just curious if this is necessary anymore.

File: cmd/codegen/generator/go/templates/modules.go [hunk 5]
```diff
@@ -824,3 +954,74 @@ func (ps *parseState) declForFunc(fnType *types.Func) (*ast.FuncDecl, error) {
 	}
 	return nil, fmt.Errorf("no decl for %s", fnType.Name())
 }
+
+// commentForPos returns the *ast* comment group for the given position. This
+// is needed because function args (despite being fields) don't have comments
+// associated with them, so this is a neat little hack to get them out.
+func (ps *parseState) commentForPos(pos token.Pos) (*ast.CommentGroup, error) {
```
Comment (sipsma) [inline cmd/codegen/generator/go/templates/modules.go:961]: 🎉

File: cmd/codegen/generator/go/templates/modules.go [hunk 6]
```diff
@@ -319,49 +356,51 @@ func (ps *parseState) fillObjectFunctionCases(type_ types.Type, cases map[string
...
 				}
 
-				fnCallArgs = append(fnCallArgs, Id(optsName))
+				target = Id(argName)
 			} else {
-				argName := strcase.ToLowerCamel(arg.Name())
+				if _, ok := vars[spec.parent.name]; !ok {
+					tp, access := findOptsAccessPattern(spec.parent.paramType, Id(spec.parent.name))
+					statements = append(statements, Var().Id(spec.parent.name).Id(renderNameOrStruct(tp)))
+					if spec.variadic {
```
Comment (sipsma) [inline cmd/codegen/generator/go/templates/modules.go:385]: nit, mostly for my own understanding: the inline struct can't be variadic, right? Just wondering if this case would ever get hit.

**Review by sipsma — APPROVED** [review]
LGTM! I left a bunch of nit-picky comments but nothing is blocking

File: cmd/codegen/generator/go/templates/modules.go (hunk 6 above)
Comment (jedevc) [reply cmd/codegen/generator/go/templates/modules.go:385]: Nope, this case isn't possible. However, the repetition in this part of the code *is* unclear, so I've refactored to have just one call to `findOptsAccessPattern` and only one check of `spec.variadic`.

File: cmd/codegen/generator/go/templates/modules.go [hunk 7]
```diff
@@ -265,9 +267,44 @@ func renderNameOrStruct(t types.Type) string {
 	if ptr, ok := t.(*types.Pointer); ok {
 		return "*" + renderNameOrStruct(ptr.Elem())
 	}
+	if sl, ok := t.(*types.Slice); ok {
```
[the remark this answers is not in the document]
Comment (jedevc) [reply cmd/codegen/generator/go/templates/modules.go:270]: I'll leave a TODO for this, and come back and grab it later :)

File: cmd/codegen/generator/go/templates/modules.go (hunk 4 above)
Comment (jedevc) [reply cmd/codegen/generator/go/templates/modules.go:901]: Aha, this isn't for object+function names, this is for the parameter names.

Even if we did the same transformation in the engine, we need to know how to get the graphql input string so we can extract the right name from the incoming request.

File: cmd/codegen/generator/go/templates/modules.go [hunk 8]
```diff
@@ -824,3 +950,113 @@ func (ps *parseState) declForFunc(fnType *types.Func) (*ast.FuncDecl, error) {
...
+func (ps *parseState) commentForFuncField(fnDecl *ast.FuncDecl, i int) (*ast.CommentGroup, error) {
+	pos := fnDecl.Type.Params.List[i].Pos()
+	tokenFile := ps.fset.File(pos)
+	if tokenFile == nil {
+		return nil, fmt.Errorf("no file for function %s", fnDecl.Name.Name)
+	}
+	line := tokenFile.Line(pos)
+
+	allowDocComment := true
+	allowLineComment := true
```
[the remark this answers is not in the document]
Comment (sipsma) [reply cmd/codegen/generator/go/templates/modules.go:966]: Ah good catch but totally agree we shouldn't block merging on this, can iterate in follow-ups

**Comment by vito** [comment]
Meant to circle back to this for a review, but thanks for taking the time to support both patterns! :pray: I looked over the changes earlier and nothing additional jumped out.
